lib/logstash/inputs/snmptrap.rb

```diff
@@ -47,26 +47,44 @@ def register
     end
   end # def register
 
-  public
   def run(output_queue)
     begin
       # snmp trap server
       snmptrap_listener(output_queue)
     rescue => e
       @logger.warn("SNMP Trap listener died", :exception => e, :backtrace => e.backtrace)
-      sleep(5)
-      retry
+      stoppable_call(20) { sleep(0.25) }
```

Comment:
specifically here with `stoppable_call` I think we should just expose a `stoppable_sleep` method, we do have various sleep calls in some plugins. this `stoppable_sleep` can be based on the same idea as done in https://github.com/colinsurprenant/ruby-stud/blob/8684e0fd8a9741035f3d9e338d0cd4b41206adcf/lib/stud/interval.rb#L64-L71